frontend/include/chpl/resolution/scope-types.h

```diff
@@ -103,51 +110,28 @@ class OwnedIdsWithName {
   reference to a collection stored in OwnedIdsWithName.
  */
 class BorrowedIdsWithName {
- private:
-  // TODO: consider storing a variant of ID here
-  // with symbolPath, postOrderId, and tag
-  ID id_;
-  const std::vector<ID>* moreIds_ = nullptr;
- public:
-  /** Construct an empty BorrowedIdsWithName */
-  BorrowedIdsWithName() { }
-  /** Construct a BorrowedIdsWithName referring to one ID */
-  BorrowedIdsWithName(ID id) : id_(std::move(id)) { }
-  /** Construct a BorrowedIdsWithName referring to the same IDs
-      as the passed OwnedIdsWithName.
-      This BorrowedIdsWithName assumes that the OwnedIdsWithName
-      will continue to exist. */
-  BorrowedIdsWithName(const OwnedIdsWithName& o)
-    : id_(o.id_), moreIds_(o.moreIds_.get())
-  { }
+ // To allow us to use the private constructor
+ friend class OwnedIdsWithName;
 
-  /** Return the number of IDs stored here */
-  int numIds() const {
-    if (moreIds_ == nullptr) {
-      return 1;
-    }
-    return moreIds_->size();
+ private:
+  static inline bool isIDVisible(const OwnedID& id, bool skipPrivateVisibilities) {
+    return !skipPrivateVisibilities || id.second != uast::Decl::PRIVATE;
   }
 
-  /** Returns the i'th ID. id(0) is always available. */
-  const ID& id(int i) const {
-    if (i == 0) {
-      return id_;
-    }
-    assert(moreIds_ && 0 <= i && (size_t) i < moreIds_->size());
-    return (*moreIds_)[i];
+  bool isIDVisible(const OwnedID& id) const {
+    return isIDVisible(id, skipPrivateVisibilities);
   }
 
   /** Returns an iterator referring to the first element stored. */
-  const ID* begin() const {
+  const OwnedID* beginOwned() const {
```

Comment:
If we drop the name `OwnedID`, is it fine to just leave this as `begin` and `end`?

Reply:
No, since `begin` and `end` are used with a different return type to actually provide C++11-style for loops.

Reply:
Thanks for the reminder, good point.

Reply:
Can we rename it to `beginIdAndVis` then?

Reply:
Yep, sure thing.